File: algorithms/23_elixir/sorting/bubble_sort.ex
defmodule BubbleSort do
  def sort([]), do: []
  def sort(list) do
    {new_list, swapped?} = pass(list)
    if swapped?, do: sort(new_list), else: new_list
  end
  defp pass([a, b | rest]) when a > b, do: {[b | elem(pass([a | rest]), 0)], true}
  defp pass([a | rest]) do
    {sorted, s} = pass(rest)
    {[a | sorted], s}
  end
  defp pass([]), do: {[], false}
end

File: algorithms/22_haskell/sorting/BubbleSort.hs
module BubbleSort (bubbleSort) where
bubbleSort :: Ord a => [a] -> [a]
bubbleSort [] = []
bubbleSort xs = let (xs', swapped) = pass xs
                in if swapped then bubbleSort xs' else xs'
  where
    pass [] = ([], False)
    pass [x] = ([x], False)
    pass (x:y:rest)
      | x > y = let (r,_) = pass (x:rest) in (y:r, True)
      | otherwise = let (r,s) = pass (y:rest) in (x:r, s)